kubernetes/resource_kubernetes_job.go

```diff
@@ -73,6 +79,15 @@ func resourceKubernetesJobCreate(d *schema.ResourceData, meta interface{}) error
 
 	d.SetId(buildId(out.ObjectMeta))
 
+	namespace, name, err := idParts(d.Id())
+	if err != nil {
+		return err
+	}
+	if d.Get("wait_for_completion").(bool) {
+		return resource.Retry(d.Timeout(schema.TimeoutCreate),
+			waitForJobCondition(conn, namespace, name, batchv1.JobComplete))
```

Comment:
Looking at this a bit more - if we run a job that fails immediately, Terraform will just wait it out and eventually error when the timeout is reached. The error message is not super helpful: 

```
Error: job: default/job-with-wait is not in: Complete condition
```

What do you think about also capturing condition type == Failed and returning a NonRetryableError telling the user that the Job failed? It will save people waiting around for a few minutes when we can tell them their job failed right away.

Reply:
Yes, I agree, current job failure handling is less than ideal. 

The `retryUntilJobCondition` initially is intended to be generic so that it can be used to wait for job failures (if needed). 
With that said, I don't see an apparent use case when a user will run a job that intended to fail (doesn't mean that it isn't a valid use-case)

What if we provide a list of conditions, i.e.
```go
func retryUntilJobHasACondition(conn *kubernetes.Clientset, ns, name string, conditions []batchv1.JobConditionType) resource.RetryFunc {
...
```
This could allow us to use it as the following:
```go
retryUntilJobCondition(conn, namespace, name, []batchv1.JobConditionType{batchv1.JobComplete, batchv1.JobFailed}))
```

Reply:
I can totally think of use-cases where you would _expect_ that something is going to fail, i.e you might use terraform to start a job that runs some kind of test as part of your CI and then report on it having failed. 

I was thinking more for users that might be running Jobs that contain initialization or migration steps or something like that - where it would be helpful for the user for Terraform to explicitly tell them that the job failed, rather than just letting it time out - it takes a way a little bit of uncertainty. The set of reasons for a Job failing is a lot smaller than for a Job having timed out.

The issue with just accepting a list of conditions is that the resource will say it was created successfully even if the job ends up with a failed condition. This might confuse users who use a Job as part of a `depends_on`.

What about supplying two lists of conditions, one list of "success" conditions that will consider the resource created, and another list of "failure" conditions that will bubble up as an error to the user?

Reply:
> The set of reasons for a Job failing is a lot smaller than for a Job having timed out.

I think you made a compelling argument and I am coming around on:

> What do you think about also capturing condition type == Failed and returning a NonRetryableError telling the user that the Job failed? It will save people waiting around for a few minutes when we can tell them their job failed right away.

Reply:
Thank you for the review and feedback, @jrhouston. 

I incorporated most of the feedback suggestions, except the:
> An idea that springs to mind having the test create a job that takes duration N to run, and asserting that the resource took no less than duration N to be created.

It has been a while since I initially drafted this PR (back in Sep) and TBH even back then I think I didn't understand well enough terraform provider unit tests, so I may need some additional help with that.

Reply:
No problem, thanks for making those changes! I figured out a way doing this so I will add a commit to your branch with the test then we can merge.